internal/app/backend/backend_service.go

```diff
@@ -71,38 +71,65 @@ func (s *backendService) FetchMatches(req *pb.FetchMatchesRequest, stream pb.Bac
 		return status.Errorf(codes.InvalidArgument, "match function configuration needs to be provided")
 	}
 
-	ctx := stream.Context()
-	var c <-chan *pb.Match
+	var grpcClient pb.MatchFunctionClient
+	// TODO: Uncomment when rest config is implemented
+	// var httpClient *http.Client
+	// var baseURL string
+	var err error
+
 	switch (req.Config.Type).(type) {
 	// MatchFunction Hosted as a GRPC service
 	case *pb.FunctionConfig_Grpc:
-		client, err := s.getGRPCClient((req.Config.Type).(*pb.FunctionConfig_Grpc))
+		grpcClient, err = s.getGRPCClient((req.Config.Type).(*pb.FunctionConfig_Grpc))
 		if err != nil {
 			logger.WithFields(logrus.Fields{
 				"error":    err.Error(),
 				"function": req.Config,
 			}).Error("failed to connect to match function")
 			return status.Error(codes.InvalidArgument, "failed to connect to match function")
 		}
-
-		// Get the channel over which the generated match results will be sent.
-		c = s.matchesFromGrpcMMF(ctx, req.Profile, client)
-
 	// MatchFunction Hosted as a REST service
 	case *pb.FunctionConfig_Rest:
+		// httpClient, baseURL, err := s.getHTTPClient((req.Config.Type).(*pb.FunctionConfig_Rest))
 		return status.Error(codes.Unimplemented, "not implemented")
-
 	default:
 		logger.Error("unsupported function type provided")
 		return status.Error(codes.InvalidArgument, "provided match function type is not supported")
 	}
 
+	ctx := stream.Context()
+	matchChan := make(chan *pb.Match)
+	errChan := make(chan error)
+
+	go func(matchChan chan<- *pb.Match, errChan chan<- error) {
+		var wg sync.WaitGroup
+		for _, profile := range req.Profile {
+			wg.Add(1)
+			go func(profile *pb.MatchProfile) {
+				defer wg.Done()
+
+				switch (req.Config.Type).(type) {
+				case *pb.FunctionConfig_Grpc:
+					// Get the channel over which the generated match results will be sent.
+					s.matchesFromGrpcMMF(ctx, profile, grpcClient, matchChan, errChan)
+				case *pb.FunctionConfig_Rest:
+					// TODO: implement matchesFromHttpMMF function
+					// s.matchesFromHttpMMF(ctx, profile, httpClient, baseURL, matchChan, errChan)
+				}
+			}(profile)
+		}
+		wg.Wait()
+		close(matchChan)
```

Comment:
defer func() {
  wg.Wait()
  close(matchChan)
  close(errChan)
}()
At the top of this func. Otherwise there's a decent chance of leaking these resources.

Also keep the wg.Wait() here as well.